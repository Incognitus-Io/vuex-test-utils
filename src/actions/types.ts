import { DispatchOptions, CommitOptions, ActionContext } from 'vuex';

declare global {
    namespace Chai {
        export interface ExpectStatic {
            action<S, R>(
                action: actionFn<S, R>,
                payload?: any,
                config?: ActionCtx<S, R>,
            ): Assertion | PromisedAssertion;
        }

        interface Assertion {
            commit: VuexCommits;
            dispatch: VuexDispatch;
            getAwaiter: PromiseLike<any>;
        }

        interface PromisedAssertion extends PromiseLike<any> {
            commit: VuexCommits;
            dispatch: VuexDispatch;
            getAwaiter: PromiseLike<any>;
        }

        interface VuexAssertion extends Assertion {
            and: VuexAssertion;
            as: VuexAssertion;
            is: VuexAssertion;
            not: VuexAssertion;
            root: VuexAssertion;
            partially: VuexActionCtxAssertions;
            containing: VuexContaining;
        }

        interface VuexCommitAssertions extends VuexAssertion {
            is: VuexCommitAssertions;
            not: VuexCommitAssertions;
            silent: VuexCommitAssertions;
        }

        interface VuexActionCtxAssertions {
            in: VuexOrder;
            partially: VuexActionCtxAssertions;
            containing: VuexContaining;
            getAwaiter: PromiseLike<any>;
        }

        interface VuexCommits extends VuexActionCtxAssertions {
            (type?: string): VuexCommitAssertions;
        }

        interface VuexDispatch extends VuexActionCtxAssertions {
            (type: string): VuexAssertion;
        }

        interface VuexContaining {
            payload: (payload: any) => Assertion;
        }

        interface VuexOrder {
            order: (...types: string[]) => Assertion;
        }

        interface ActionCtx<S, R> {
            state?: S;
            rootState?: R;
            getters?: any;
            rootGetters?: any;
        }
    }

}

export interface ObservedBase {
    type: string;
    payload?: any;
}
export interface ObservedDispatch extends ObservedBase {
    options?: DispatchOptions;
}
export interface ObservedCommit extends ObservedBase {
    options?: CommitOptions;
}
export type commitFn = (_: string, __?: any, ___?: CommitOptions) => void;
export type dispatchFn = (_: string, __?: any, ___?: DispatchOptions) => any;
export type actionFn<S, R> = (_: ActionContext<S, R>, __?: any) => void | Promise<void>;
export type actionFnSync<S, R> = (_: ActionContext<S, R>, __?: any) => void;
export type actionMode = 'commit' | 'dispatch';
