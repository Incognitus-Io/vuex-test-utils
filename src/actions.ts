import { ActionContext, CommitOptions, DispatchOptions } from 'vuex';
import { ActionResults, ObservedCommit, ObservedDispatch, actionFn, commitFn, dispatchFn, ActionCtx } from './index';

export const action = <S, R>(
    action: actionFn<S, R>,
    payload?: any,
    config?: ActionCtx<S, R>): jest.JestMatchers<ActionResults> => {
    const executedCommits: ObservedCommit[] = [];
    const executedDispatches: ObservedDispatch[] = [];
    const ctx = config as ActionContext<S, R> || {}
    ctx.commit = emitCommit(executedCommits);
    ctx.dispatch = emitDispatch(executedDispatches);

    action(ctx, payload);

    return expect({
        commits: executedCommits,
        dispatches: executedDispatches,
    });
};

const emitCommit = (executedCommits: ObservedCommit[]): commitFn => {
    return (type: string, payload?: any, options?: CommitOptions) => {
        executedCommits.push({
            type,
            payload,
            options,
        });
    };
};

const emitDispatch = (executedDispatches: ObservedDispatch[]): dispatchFn => {
    return (type: string, payload?: any, options?: DispatchOptions) => {
        executedDispatches.push({
            type,
            payload,
            options,
        });
    };
};