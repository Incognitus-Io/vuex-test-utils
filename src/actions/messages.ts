export const messages = {
    expected: {
        commitedType: 'expected #{exp} commit but found #{act} commit(s)',
        orderedCommitedType: 'expected #{exp} commit but found #{act}',
        payload: 'expected payload #{exp} but found #{act}',
        missingRootCommitOptions: 'expected to be a root commit, but found no commit options',
        missingSilentCommitOptions: 'expected to be a silent commit, but found no commit options',
        missingRootDispatchOptions: 'expected to be a root dispatch, but found no dispatch options',
        dispatchType: 'expected #{exp} dispatch but found #{act} dispatch(s)',
        orderedDispatchType: 'expected #{exp} dispatch but found #{act}',
    },
    notExpected: {
        commitedType: 'expected #{exp} to not be commmited but found #{act} commit(s)',
        orderedCommitedType: 'expected #{exp} to not be commited but found #{act}',
        payload: 'expected payload #{exp} to not be defined but found #{act}',
        dispatchType: 'expected #{exp} to not be dispatched but found #{act} dispatches(s)',
        orderedDispatchType: 'expected #{exp} to not be dispatched but found #{act}',
    },
};
