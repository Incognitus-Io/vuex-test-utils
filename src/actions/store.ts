export const store = {
    not: 'negate',
    notAsync: 'negateAsync',
    executedCommits: 'executedCommits',
    currentCommit: 'currentCommit',
    executedDispatches: 'executedDispatches',
    currentDispatch: 'currentDispatch',
    actionPayload: 'actionPayload',
    actionCtx: 'actionContext',
    actionMode: 'actionMode',
    partially: 'partially',
};
