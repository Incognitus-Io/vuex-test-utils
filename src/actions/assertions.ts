import { ObservedBase, actionMode, ObservedCommit, ObservedDispatch } from './types';
import { store } from './store';
import { messages } from './messages';
import { AssertionError } from 'chai';

export class Assertions {
    private Assertion: Chai.AssertionStatic;
    private flag: (obj: object, key: string, value?: any) => any;
    private eql: (expected: any, actual?: any) => boolean;

    constructor(assertion: Chai.AssertionStatic, utils: Chai.ChaiUtils) {
        this.Assertion = assertion;
        this.flag = utils.flag;
        this.eql = utils.eql;
    }

    public order(that: Chai.AssertionStatic, ...types: string[]) {
        let executed: ObservedBase[];
        let failMessage: string;
        let negFailMessage: string;
        const mode: actionMode = this.flag(that, store.actionMode);

        if (mode === 'commit') {
            executed = this.flag(that, store.executedCommits);
            failMessage = messages.expected.orderedCommitedType;
            negFailMessage = messages.notExpected.orderedCommitedType;
        } else {
            executed = this.flag(that, store.executedDispatches);
            failMessage = messages.expected.orderedDispatchType;
            negFailMessage = messages.notExpected.orderedDispatchType;
        }

        let executedTypes = executed.map((x) => x.type);

        const firstCommitIdx = executedTypes.indexOf(types[0]);
        executedTypes = executedTypes.splice(firstCommitIdx, types.length);

        executedTypes.forEach((type, idx) => that.assert(
            type === types[idx],
            failMessage, negFailMessage,
            types[idx], type,
            true,
        ));
    }

    public commit(that: Chai.AssertionStatic, type: string) {
        const executedCommits: ObservedCommit[] = this.flag(that, store.executedCommits);
        const executedTypes = executedCommits.map((x) => x.type);

        const currentCommitIdx = executedTypes.indexOf(type);
        const results = currentCommitIdx !== -1;

        if (results) {
            const currentCommit = executedCommits[currentCommitIdx];
            this.flag(that, store.currentCommit, currentCommit);

            executedCommits.splice(currentCommitIdx, 1);
            this.flag(that, store.executedCommits, executedCommits);
        }

        that.assert(
            results,
            messages.expected.commitedType, messages.notExpected.commitedType,
            type, executedTypes.join(', '),
        );
    }

    public dispatch(that: Chai.AssertionStatic, type: string) {
        const executedDispatches: ObservedDispatch[] = this.flag(that, store.executedDispatches);
        const executedTypes = executedDispatches.map((x) => x.type);

        const currentDispatchIdx = executedTypes.indexOf(type);
        const results = currentDispatchIdx !== -1;

        if (results) {
            const currentDispatch = executedDispatches[currentDispatchIdx];
            this.flag(that, store.currentDispatch, currentDispatch);

            executedDispatches.splice(currentDispatchIdx, 1);
            this.flag(that, store.executedDispatches, executedDispatches);
        }

        that.assert(
            results,
            messages.expected.dispatchType, messages.notExpected.dispatchType,
            type, executedTypes.join(', '),
        );
    }

    public payload(that: Chai.AssertionStatic, payload: any) {
        let current: ObservedBase;
        let executed: ObservedBase[];
        const mode: actionMode = this.flag(that, store.actionMode);

        if (mode === 'commit') {
            current = this.flag(that, store.currentCommit);
            executed = (!current)
                ? this.flag(that, store.executedCommits)
                : [current];
        } else {
            current = this.flag(that, store.currentDispatch);
            executed = (!current)
                ? this.flag(that, store.executedDispatches)
                : [current];
        }

        const executedPayloads = executed.map((x) => x.payload);
        const results = executedPayloads.some((x) => this.eql(payload, x));
        that.assert(
            results,
            messages.expected.payload, messages.notExpected.payload,
            payload, executedPayloads,
        );
    }

    public root(that: Chai.AssertionStatic) {
        let current: ObservedCommit | ObservedDispatch;
        let msg: string;
        const mode: actionMode = this.flag(that, store.actionMode);
        const negated = this.flag(that, store.not) || false;

        if (mode === 'commit') {
            current = this.flag(that, store.currentCommit);
            msg = messages.expected.missingRootCommitOptions;
        } else {
            current = this.flag(that, store.currentDispatch);
            msg = messages.expected.missingRootDispatchOptions;
        }

        if (!current.options) {
            throw new AssertionError(msg);
        }

        const test = new this.Assertion(current.options.root).to.be;
        // tslint:disable-next-line:no-unused-expression
        (negated) ? test.false : test.true;
    }

    public silent(that: Chai.AssertionStatic) {
        const currentCommit: ObservedCommit = this.flag(that, store.currentCommit);
        const negated = this.flag(that, store.not) || false;

        if (!currentCommit.options) {
            throw new AssertionError(messages.expected.missingSilentCommitOptions);
        }

        const test = new this.Assertion(currentCommit.options.silent).to.be;
        // tslint:disable-next-line:no-unused-expression
        (negated) ? test.false : test.true;
    }
}
