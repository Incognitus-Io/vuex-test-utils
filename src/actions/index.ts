import { ActionContext, CommitOptions, DispatchOptions } from 'vuex';
// import './types';
import chaiEx from './register';
import { actionFn, ObservedCommit, ObservedDispatch, commitFn, dispatchFn } from './types';
import { store } from './store';

export const vuexChai = (chai: Chai.ChaiStatic, _: Chai.ChaiUtils) => {
    const Assertion = chai.Assertion;
    chaiEx(chai, _);

    chai.expect.action = <S, R>(
        action: actionFn<S, R>,
        payload?: any,
        config?: Chai.ActionCtx<S, R>,
    ): Chai.Assertion  => {
        const test = new Assertion(action);
        _.flag(test, store.actionCtx, config);

        execAction(action, payload, test);

        return test;
    };

    const execAction = (action: actionFn<any, any>, payload: any, obj: object) => {
        const executedCommits: ObservedCommit[] = [];
        const executedDispatches: ObservedDispatch[] = [];
        const ctx: ActionContext<any, any> = _.flag(obj, store.actionCtx) || {};
        ctx.commit = emitCommit(executedCommits);
        ctx.dispatch = emitDispatch(executedDispatches);

        action(ctx, payload);

        _.flag(obj, store.executedCommits, executedCommits);
        _.flag(obj, store.executedDispatches, executedDispatches);
    };

    const emitCommit = (executedCommits: ObservedCommit[]): commitFn => {
        return (type: string, payload?: any, options?: CommitOptions) => {
            executedCommits.push({
                type,
                payload,
                options,
            });
        };
    };

    const emitDispatch = (executedDispatches: ObservedDispatch[]): dispatchFn => {
        return (type: string, payload?: any, options?: DispatchOptions) => {
            executedDispatches.push({
                type,
                payload,
                options,
            });
        };
    };
};
