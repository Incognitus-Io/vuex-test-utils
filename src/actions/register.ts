import { store } from './store';
import { Assertions as VuexTestUtilsAssertions } from './assertions';

export const chaiExtensions = (chai: Chai.ChaiStatic, _: Chai.ChaiUtils) => {
    const Assertion = chai.Assertion;
    const VuexAssertions = new VuexTestUtilsAssertions(Assertion, _);

    const isPromise = (obj: any) => obj && typeof obj.then === 'function';

    [
        nameof<Chai.VuexCommits>((x) => x.in),
        nameof<Chai.VuexCommits>((x) => x.containing),
        nameof<Chai.VuexAssertion>((x) => x.as),
    ].forEach((chain) => {
        Assertion.addProperty(chain, function () {
            if (isPromise(this._obj)) {
                return this;
            }
        });
    });

    Assertion.addProperty(nameof<Chai.Assertion>((x) => x.getAwaiter), function () {
        return this._obj as PromiseLike<any>;
    });

    Assertion.addProperty(nameof<Chai.VuexActionCtxAssertions>((x) => x.partially), function () {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => {
                _.flag(test, store.partially, true);
                _.flag(test, store.deep, true);
                return test; });
            return this;
        } else {
            _.flag(this, store.partially, true);
        }
    });

    Assertion.addMethod(nameof<Chai.VuexOrder>((x) => x.order), function (...types: string[]) {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => VuexAssertions.order(test, ...types));
            return this;
        } else {
            VuexAssertions.order(this, ...types);
        }
    });

    Assertion.addChainableMethod(nameof<Chai.Assertion>((x) => x.commit), function (type?: string) {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            if (type) {
                this._obj = promise.then((test) => VuexAssertions.commit(test, type));
            }
            return this;
        } else {
            VuexAssertions.commit(this, type!);
        }
    }, function () {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => {
                _.transferFlags(this as any, test);
                _.flag(test, store.actionMode, 'commit');
                _.flag(test, store.not, _.flag(test, store.notAsync));
                return test;
            });
        } else {
            _.flag(this, store.actionMode, 'commit');
        }
    });

    Assertion.addChainableMethod(nameof<Chai.Assertion>((x) => x.dispatch), function (type: string) {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => VuexAssertions.dispatch(test, type));
            return this;
        } else {
            VuexAssertions.dispatch(this, type);
        }
    }, function () {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => {
                _.transferFlags(this as any, test);
                _.flag(test, store.actionMode, 'dispatch');
                _.flag(test, store.not, _.flag(test, store.notAsync));
                return test;
            });
        } else {
            _.flag(this, store.actionMode, 'dispatch');
        }
    });

    Assertion.addMethod(nameof<Chai.VuexContaining>((x) => x.payload), function (payload: any) {

        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => VuexAssertions.payload(test, payload));
            return this;
        } else {
            VuexAssertions.payload(this, payload);
        }
    });

    Assertion.addProperty(nameof<Chai.VuexAssertion>((x) => x.root), function () {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => VuexAssertions.root(test));
            return this;
        } else {
            VuexAssertions.root(this);
        }
    });

    Assertion.addProperty(nameof<Chai.VuexCommitAssertions>((x) => x.silent), function () {
        const promise = getPromiseOrDefault(this);
        if (promise) {
            this._obj = promise.then((test) => VuexAssertions.silent(test));
            return this;
        } else {
            VuexAssertions.silent(this);
        }
    });

    const getPromiseOrDefault = (assertion: Chai.AssertionStatic): PromiseLike<any> | undefined =>
        (isPromise(assertion._obj)) ? assertion._obj as PromiseLike<any> : undefined;
};

export default chaiExtensions;
