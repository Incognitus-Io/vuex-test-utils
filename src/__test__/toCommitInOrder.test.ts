import { ActionTree, ActionContext } from 'vuex';

import { action } from '../actions';
import matcher from '../toCommitInOrder';

expect.extend(matcher)

interface SyncAction extends ActionTree<any, any> {
    foobar(ctx: ActionContext<any, any>): void;
}

describe('.toCommitInOrder', () => {
    it('passes when the action commits mutations', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        action(actions.foobar).toCommitInOrder(['1', '2', '3']);
    });
    it('passes when the acton commits some mutations when not strict', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        action(actions.foobar).toCommitInOrder(['1', '3'], false);
    });
    it('fails when the action does not commits mutations', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        expect(() => action(actions.foobar)
            .toCommitInOrder(['1', '2'])
        ).toThrowErrorMatchingSnapshot();
    });
});

describe('.not.toCommitInOrder', () => {
    it('passes when the action does not commits mutations', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        action(actions.foobar).not.toCommitInOrder(['1', '3', '2']);
    });
    it('fails when the action commits mutations not strictly', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        expect(() => action(actions.foobar)
            .not.toCommitInOrder(['1', '3'], false)
        ).toThrowErrorMatchingSnapshot();
    });
    it('fails when the action commits mutations', () => {
        const actions = {
            foobar: (ctx) => {
                ctx.commit('1');
                ctx.commit('2');
                ctx.commit('3');
            }
        } as SyncAction;

        expect(() => action(actions.foobar)
            .not.toCommitInOrder(['1', '2', '3'])
        ).toThrowErrorMatchingSnapshot();
    });
});
