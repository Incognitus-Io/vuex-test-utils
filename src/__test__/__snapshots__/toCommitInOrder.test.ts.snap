// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.not.toCommitInOrder fails when the action commits mutations 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toCommitInOrder()[22m

Expected to not have been commited:
 [32m[\\"1\\", \\"2\\", \\"3\\"][39m
Received:
 [31m[\\"1\\", \\"2\\", \\"3\\"][39m"
`;

exports[`.not.toCommitInOrder fails when the action commits mutations not strictly 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).not.toCommitInOrder()[22m

Expected to not have been commited:
 [32m[\\"1\\", \\"3\\"][39m
Received:
 [31m[\\"1\\", \\"2\\", \\"3\\"][39m"
`;

exports[`.toCommitInOrder fails when the action does not commits mutations 1`] = `
"[2mexpect([22m[31mreceived[39m[2m).toCommitInOrder()[22m

Expected to have been commit:
 [32m[\\"1\\", \\"2\\"][39m
Received:
 [31m[\\"1\\", \\"2\\", \\"3\\"][39m"
`;
